### PR TITLE
feat: allow using authorization query parameter

### LINKED ISSUE #387: Authentication with multiple JWT tokens from the same browser session
I have a question about serving multi-tenant / multi-client apps.

We have a setup like this one:
* a single multi-tenant API Platform backend (on `paas.org`)
* a single Mercure hub, which is on the same domain as PaaS
* a multi-client-enabled CMS (on `tenant.org` domain)
  * the CMS is allowed to administrate all the clients for their tenant 
* multiple clients (websites) are on their own domains, too (`client-one.org`, `client-two.org`...)

_For reference:
in this setup the **tenants** (our customers) are parents of **clients** (our customers have multiple websites, mobile apps, etc.)._

When working just with a single client, we ran into the problem that the client and the hub are not on the same second-level domain. I've seen #17 and the polyfill solution is not what we would want since that falls back to plain ol' HTTP and as such is not as "native" and performant as the real native `EventSource`.
We could/should fix this with the PaaS setting the `mercureAuthentication` cookie at the time of response to OAuth, and since the hub is on the same domain this should work... almost, I'm not going into detail with OAuth scopes. But still, even that is not the whole story. :slightly_smiling_face: 

When switching to a second client in the CMS, it has to fetch a different JWT from PaaS (which would reset the cookie), then the CMS opens a new `EventSource` connection to the hub with the new JWT.
With the solution when the PaaS sets the mercure cookie, we here _loose_ the `EventSource` connection for the first client.
If we exposed the header to JS, it still wouldn't work because JS can not set a cookie for another domain (it's not a subdomain, so a `Domain` attribute isn't suitable either).

We want to enable the customer to have multiple tabs with CMS open in the same browser session, each administering it's own client. With cookie-based mercure authentication we can not have each tab receiving it's own SSEs because the cookie can only be set to a single value.

I've found this #150 and I understand the goal of securing the JWT. But in the end, with HTTPS the URI itself is encrypted in transport. I'd suggest to add a query parameter as a possibility for mercure authentication. The current limitation is restrictive enough so we're not able to accomplish what we've aimed for.

I'm open to any other suggestions which would resolve this problem, without the need to transport the JWT in a query parameter. Please advise!

### PR DESCRIPTION
**Pull request by azjezz**
closes #387

### CODE REVIEW

File: authorization.go [hunk 1]
```diff
@@ -58,6 +60,15 @@ func authorize(r *http.Request, jwtConfig *jwtConfig, publishOrigins []string, c
 		return validateJWT(authorizationHeaders[0][7:], jwtConfig)
 	}
 
+	authorizationQuery, queryExists := r.URL.Query()["authorization"]
```
Comment (dunglas) [inline authorization.go:63]: You could probably merge with this line with the next one.

**Review by dunglas — APPROVED** [review]

File: authorization.go (hunk 1 above)
Comment (azjezz) [reply authorization.go]: done.

**Comment by dunglas** [comment]
Could you also add some config in `Caddyfile` to strip the value of the `authorization` query parameter from the logs? https://dunglas.fr/2022/04/caddy-logging-security-improvements/

File: authorization_test.go [hunk 2]
```diff
@@ -207,6 +207,133 @@ func TestAuthorizeAuthorizationHeaderWrongAlgorithm(t *testing.T) {
...
+	claims, err := authorize(r, &jwtConfig{[]byte{}, jwt.SigningMethodHS256}, []string{}, defaultCookieName)
+	assert.EqualError(t, err, "unable to parse JWT: signature is invalid")
+	assert.Nil(t, claims)
+}
+
+func TestAuthorizeAuthorizationQueryInvalidKeyRsa(t *testing.T) {
+	r, _ := http.NewRequest("GET", defaultHubURL, nil)
+	query := r.URL.Query()
+	query.Set("authorization", validEmptyHeaderRsa)
+	r.URL.RawQuery = query.Encode()
```
Comment (azjezz) [inline authorization_test.go:247]: having to do this over and over is kinda of annoying :)

**Comment by azjezz** [comment]
tests are passing 👍

**Review by dunglas — APPROVED** [review]

**Comment by dunglas** [comment]
Thanks @azjezz

**Pull request merged**